packages/go_router/lib/src/builder.dart

```diff
@@ -178,78 +207,110 @@ class RouteBuilder {
-      _buildRecursive(context, matchList, startIndex + 1, onPopPage,
+      _buildRecursive(context, matchList, startIndex + 1, pagePopContext,
           routerNeglect, keyToPages, navigatorKey, registry);
-    } else if (route is ShellRoute) {
+    } else if (route is ShellRouteBase) {
+      assert(startIndex + 1 < matchList.matches.length,
+          'Shell routes must always have child routes');
       // The key for the Navigator that will display this ShellRoute's page.
       final GlobalKey<NavigatorState> parentNavigatorKey = navigatorKey;
 
-      // The key to provide to the ShellRoute's Navigator.
-      final GlobalKey<NavigatorState> shellNavigatorKey = route.navigatorKey;
-
-      // The observers list for the ShellRoute's Navigator.
-      final List<NavigatorObserver> observers =
-          route.observers ?? <NavigatorObserver>[];
-
       // Add an entry for the parent navigator if none exists.
       keyToPages.putIfAbsent(parentNavigatorKey, () => <Page<Object?>>[]);
 
-      // Add an entry for the shell route's navigator
-      keyToPages.putIfAbsent(shellNavigatorKey, () => <Page<Object?>>[]);
-
       // Calling _buildRecursive can result in adding pages to the
       // parentNavigatorKey entry's list. Store the current length so
       // that the page for this ShellRoute is placed at the right index.
       final int shellPageIdx = keyToPages[parentNavigatorKey]!.length;
 
+      // Get the current sub-route of this shell route from the match list.
+      final RouteBase subRoute = matchList.matches[startIndex + 1].route;
+
+      // The key to provide to the shell route's Navigator.
+      final GlobalKey<NavigatorState> shellNavigatorKey =
+          route.navigatorKeyForSubRoute(subRoute);
+
+      // Add an entry for the shell route's navigator
+      keyToPages.putIfAbsent(shellNavigatorKey, () => <Page<Object?>>[]);
+
       // Build the remaining pages
-      _buildRecursive(context, matchList, startIndex + 1, onPopPage,
+      _buildRecursive(context, matchList, startIndex + 1, pagePopContext,
           routerNeglect, keyToPages, shellNavigatorKey, registry);
 
       final HeroController heroController = _goHeroCache.putIfAbsent(
           shellNavigatorKey, () => _getHeroController(context));
-      // Build the Navigator
-      final Widget child = HeroControllerScope(
-        controller: heroController,
-        child: _buildNavigator(
-            onPopPage, keyToPages[shellNavigatorKey]!, shellNavigatorKey,
-            observers: observers),
+
+      // Build the Navigator for this shell route
+      Widget buildShellNavigator(
+          List<NavigatorObserver>? observers, String? restorationScopeId) {
+        return _buildNavigator(
+          pagePopContext.onPopPage,
+          keyToPages[shellNavigatorKey]!,
+          shellNavigatorKey,
+          observers: observers ?? const <NavigatorObserver>[],
+          restorationScopeId: restorationScopeId,
+          heroController: heroController,
```

Comment:
If you mean heroController, it would be a bit cumbersome to pass it in observers, since it's used in _buildNavigator to wrap the Navigator in an HeroControllerScope, if not null. 

The only change from the main branch here is really just that the creation of the HeroControllerScope has been moved to _buildNavigator.